trailmap/pathway_analysis.py

```diff
@@ -0,0 +1,319 @@
+    rolled = ()
+    # find the shared node that appears last in the path
+    for node in path[-1:0:-1]:
+        if node in cycle:
+            rolled = tuple(np.roll(cycle, -cycle.index(node)),)
+            break
+
+    return rolled
+
+
+def insert_cycles(path, rolled_cycles):
+    '''Inserts already-rolled cycles as a tuple into a path
+    '''
+    path=list(path)
+    for rc in rolled_cycles:
+        path.insert(path.index(rc[0]), rc)
+    path_with_cycles = tuple(path)
+    return path_with_cycles
+
+
+def get_pathways_with_cycles(pathways, sc):
+    '''For the given list of pathways and simple cycles, returns a set of
+    pathways that could contain 1 or more cycles
+    '''
+    pathways_with_cycles = set()
+    pathways = list(pathways) # turn into list, which supports indexing
+    # look at all pathways and simple cycles
+    for path in pathways:
+        rolled_cycles = set()
+        for cycle in sc:
+            rolled_cycle = roll_cycle(path, cycle)
+            if rolled_cycle:
+                rolled_cycles.add(rolled_cycle)
+        
+        # record all the pathways that have cycles, insert single cycle
+        if rolled_cycles:
+            pathways_with_cycles.add(insert_cycles(path, rolled_cycles))
+
+    return pathways_with_cycles
+
+
+def find_paths_with_source(pathways, source):
+    '''returns a subset of pathways that contain a given facility as the source
+    '''
+    subset_pathways = set()
+    for path in pathways:
+        if path[0] == source:
+            subset_pathways.add(path)
+
+    return subset_pathways
+
+
+def find_paths_with_sink(pathways, sink):
+    '''returns a subset of pathways that contain a given facility as the sink
+    '''
+    subset_pathways = set()
+    for path in pathways:
+        if path[-1] == sink:
+            subset_pathways.add(path)
+
+    return subset_pathways
+
+
+def find_paths_containing_all(pathways, facilities):
+    '''returns a subset of pathways that contain all facilities in input list
+    '''
+    # convert to list if user passed a string or int
+    if type(facilities) == int or type(facilities) == str:
+        facilities = [facilities]
+    
+    # if user passed an empty list, return no pathways
+    if not facilities:
+        return set()
+
+    p = set()
+    for path in pathways:
+        if set(facilities).issubset(path):
+            p.add(path)
+
+    return p
+
+
+def find_paths_containing_one_of(pathways, facilities):
+    '''Returns a subset of pathways that contain one or more facilities in
+    input list
+    '''
+    # convert to list if user passed a string or int
+    if type(facilities) == int or type(facilities) == str:
+        facilities = [facilities]
+    
+    # if user passed an empty list, return no pathways
+    if not facilities:
+        return set()
+
+    p = set()
+    for path in pathways:
+        contains = [path for facility in facilities if facility in path]
+        if contains:
+            p.add(contains[0])
```

Comment:
would set intersection make more sense here?
```suggestion
    facilities = set(facilities)
    p = set( [ path for path in pathways if set(path).intersection(facilities)  ] )
```